07-routing-gossip.md

```diff
@@ -1119,7 +1120,59 @@ A->D's `update_add_htlc` message would be:
 And D->C's `update_add_htlc` would again be the same as B->C's direct payment
 above.
 
-## References
+# Onion Messages
+
+Onion messages allow peers to use existing connections to query for
+invoices (see [BOLT 12](12-offer-encoding.md)).  Like gossip messages,
+they are not associated with a particular local channel.  Like HTLCs,
+they use [BOLT 4](04-onion-routing.md#onion-messages) protocol for
+end-to-end encryption.
+
+Onion messages are unreliable: in particular, they are designed to
+be cheap to process and require no storage to forward.  As a result,
+there is no error returned from intermediary nodes.
+
+To enable messaging via blinded paths, there is an optional `blinding`
+parameter which allows decryption of the `enctlv` field inside the
+`onionmsg`'s `onionmsg_payload`.
+
+## The `onion_message` Message
+
+1. type: 387 (`onion_message`) (`option_onion_messages`)
+2. data:
+    * [`point`:`blinding`]
+    * [`u16`:`len`]
```

Comment:
Note that this is duplicative with the bigsize for the onion length, though I'd much prefer to keep this and drop the bigsize, or maybe move this size down to there. There is zero reason we need to use a bigsize to save the one extra byte.

Reply:
From what I understood, the `onionmsg` here corresponds to the `onion_packet` type (04-onion-routing.md, line 153) that is always 1366 bytes for payment onion but may be of a different size for message onions. That's why we need to specify the size `len` here. The `onion_packet` type then contains a `hop_payload` of variable length and that's why we need the [`bigsize`:`length`] there (this `length` can't be computed from the previous `len` because of the presence of `filler` of unknown size).
The `onionmsg_payload` tlv stream only describes the content of this `hop_payload` for onion messages.

It took me many readings of the spec to reach this understanding (assuming it is correct), the relationship between the different types must be made a lot clearer.

Reply:
Ah, right, its `hop_payload` vs `hop_payloads`. This is doubly confusing because the Onion Messages section only says its an onion "with an alternate `hop_payload`" without mentioning that, also the whole onion bytes thiggy is variable-length.